### ISSUE TITLE
Style Feedback

### ISSUE THREAD

**Issue by MartinKavik** [rebo/seed_styling_app#3]
Feedback for Seed Style (https://seed-style-hooks.netlify.app/style_home).

- [ ] 0) I think I saw some typos but it should be much better than my seed-rs.org docs.. 😺 

- [ ] 1) `Getting Started` on the home page suggest to clone and use `.../rebo/seed_styling_quickstart`. I wasn't able to compile that quickstart. And why we need two Seed Style quickstarts? - it's a bit annoying to clone two repos just to try something.

- [ ] 2) https://github.com/rebo/seed-style-quickstart-basic - it could be marked as a `template` repo so we can click the button`Use this template` instead of forking (see e.g. https://github.com/seed-rs/seed-quickstart) 

- [ ] 3) `cargo make start` vs `cargo make server` & `cargo make watch` - https://github.com/seed-rs/seed-quickstart#3-prepare-your-project-for-work ; `cargo make start` is a little bit simpler but becomes annoying quickly because of restarts.

- [ ] 4) https://seed-style-hooks.netlify.app/getting_started -> `How it works`. It would be nice to include a super simple example for each a-h) point.

- [ ] 5) App panics with too general error (`Option::unwrap`) when `presets` is missing and you want to use e.g. `seed_colors::Red::No4`.

- [ ] 6) I was a bit confused because of many (glob) imports - which libraries are imported and required and where I can find `seed_style`, `seed_colors`, `theme`, etc.. and what rules should I respect to not cause panics in runtime. Especially when rust-analyzer doesn't work. Perhaps just a beginner problem, but I had an impression that there are too many moving parts.

- [ ] 7) Why we use `GlobalStyle::default()` and not `GlobalStyle::new()`? My first idea was that `GlobalStyle::default` has some predefined _default_ styles.

- [ ] 8) I don't have a clear idea how should I use the code below / how it really works:
    ```rust    
        use_themes(
            || vec![theme::theme(), seed_style::style_presets()],
            || {
                global_styles::use_themed_styles();
                themed_view(model)
            },
        )
    ```
    - Why is there a `vec!`? What can (or should) I write into it? Should I keep ordering? What ordering?
    - When it can panics?
    - What does `use_themed_styles` mean?
    - What can I (should) write into the block in the second argument?
    - Why `global_styles` are used "after" `theme::theme()`? - the code isn't corresponding with my intuition (theme styles should override global style and therefore they should be instantiated after global styles) 
    - etc.

- [ ] 9) I'm not sure how `GlobalStyle` works. My intuition tells me that it should be something like singleton, however there are two `GlobalStyle::default()` in `global_styles.rs` (in Basic quickstart). And what `.activate_init_styles` and `.activate_styles` do? (And these two calls look a bit error-prone - I can imagine I'll forget to call them - maybe we should add at least tag `#[must_use]` somewhere.

- [ ] 10) `s().` vs `S.`. It's probably the most important thing to resolve before we can integrate Seed Style into Seed. So.. why `s().` and not `S.`?

- [ ] 11) Why I had to call `.breakpoint_scale([600, 960, 1280, 1920])` (https://seed-style-hooks.netlify.app/responsive_styling)?

- [ ] 12) 
    ```rust
         Msg::WindowResized => {
            conditionally_skip_rendering::<Breakpoint, _, _, _>(|| default_bp_theme(), orders)
        }
    ```
    looks pretty cumbersome and I assume that `conditionally_skip_rendering` is imported by glob so it pollutes API surface and I don't have an idea where it's defined. Possible alternative - the first idea:
    ```rust
        Msg::WindowResized => {
            if not(S::breakpoint(default_bp_theme()).has_changed()) { orders.skip(); }
        }
    ```

- [ ] 13) It would be nice to see the code snippet corresponding to the example - https://seed-style-hooks.netlify.app/layout

- [ ] 14) There are missing links to the repo and for feedback - something like the footer in https://seed-rs.org/. Also I don't see a link to seed-rs.org.

**Comment by rebo** [comment]
Notes:

1) Yes both should point to `git clone https://github.com/rebo/seed-style-quickstart-basic` **FIXED LOCALLY WILL PUSH SHORTLY.**
2) Template  **FIXED LOCALLY WILL PUSH SHORTLY.**
3) Cargo marke start -> watch  **FIXED LOCALLY WILL PUSH SHORTLY.**
4) Examples for a->h, **FIXED LOCALLY WILL PUSH SHORTLY.**
5) Replaced error  with this ->. 
.expect("Cannot find a theme that defines a CSS Value for your theme alias, are you sure you have provided access to that theme using `use_themes(|| THEME_NAME, ||..`") . **FIXED LOCALLY WILL PUSH SHORTLY**


6) Ideally should just have a single glob import `seed_style::*` which covers everything, however there is current clashes with Seed's px/rem functions and therefore sometimes have to do `seed_style::px. Unfortunately we need a glob import due to various traits needed. 
So in summary I would consider use of anything other than seed_style::* being needed a bug.
7) Global default to new **FIXED LOCALLY WILL PUSH SHORTLY**

8) and 9) will address in a separate issue as there are a number of concerns here. There is a reason for everything but yes I agree its not clear.

10) s(). vs S. . This is to do with the builder pattern used which is currently a consuming builder pattern for style objects. I.e. it consumes `self` and returns a new style. The advantage of this pattern is that you do not need a terminating method call. I.e. it can finish with say `s().color("red")` as opposed to `s().color("red").build_style()`. The disadvantage is that when  s() has to create a brand new instance which it does.  

The previous solution of `S.` used a `mut &self` builder pattern, the problem with this is that in certain circumstances you couldn't consume the `&mut selfs` that were returned by each css method. Therefore needing a terminating `build()` method which I was hoping to avoid.  If you have a solution where I can instantiate a consuming builder with `S.` instead of `s().` I'd be happy to implement.  There is probably a way and I'm just missing it :).

11) Breakpoint scale.  So themes implement both named aliases and also scales. Often these will refer to the same values but this is not currently enforced.  Named breakpoints are used for things like `s().only_and_above(Medium)`, and breakpoint scales are used for the very useful array arguments to css_values. I.e. `s().font_size(&[px(12),px(18)])` which correspond to 12px font size on the smallest breakpoint and 18px font size on the next breakpoint and up.  

I could include this explanation in the guide or we could think of someway unifying breakpoint scales and named breakpoints.

12) `conditionally_skip_rendering` yeah agree it's a random global api call. I basically wrote it so it worked and dumped it in the glob :) . I'll make a list of the api surface separately and we can work out how to encapsulate/ minimise will put. up as separate issue.

13) yes although I'm thinking of simplifying this api so have left it unfinished for now. It works but is too verbose for my liking. TODO. 

14) Footer/info etc/ WILL FIX.

**Comment by MartinKavik** [comment]
Ad 10) s() vs S. - one idea: https://gist.github.com/MartinKavik/82503487178dfe8470ee50dec832d530

**Comment by MartinKavik** [comment]
Ad 11) Breakpoints

> we could think of someway unifying breakpoint scales and named breakpoints.

Idea n.1:
Old code:
```rust
    use MySeedBreakpoint::*;
    Theme::new("default_bp_theme")
        .set_breakpoint(ExtraSmall, (0, Some(600))) 
        .set_breakpoint(Small, (600, Some(960)))
        .set_breakpoint(Medium, (960, Some(1280)))
        .set_breakpoint(Large, (1280, Some(1920)))
        .set_breakpoint(ExtraLarge, (1920, None))
        .breakpoint_scale([600, 960, 1280, 1920]) 
```
New code:
```rust
    use Breakpoint::*;
    Theme::new("my_theme").set_breakpoints(breakpoints![
        0, ExtraSmall, 600, Small, 960, Medium, 1280, Large, 1920, ExtraLarge
    ]);
```
https://gist.github.com/MartinKavik/7bafc2d37378685b7c9dcace9b92d8c8

**Comment by rebo** [comment]
ad breakpoints >  Agreed something like that would work well

`S.font_size(..)` pattern is a possibility, I will review this.

**Comment by rebo** [comment]
> Feedback for Seed Style (https://seed-style-hooks.netlify.app/style_home).
> 
> * [ ]  0) I think I saw some typos but it should be much better than my seed-rs.org docs.. 😺

TODO, Will move to new issue so can close most of these

> 
> * [ ]  1) `Getting Started` on the home page suggest to clone and use `.../rebo/seed_styling_quickstart`. I wasn't able to compile that quickstart. And why we need two Seed Style quickstarts? - it's a bit annoying to clone two repos just to try something.
> 

Fixed on main site.

> * [ ]  2) https://github.com/rebo/seed-style-quickstart-basic - it could be marked as a `template` repo so we can click the button`Use this template` instead of forking (see e.g. https://github.com/seed-rs/seed-quickstart)

Fixed 

> 
> * [ ]  3) `cargo make start` vs `cargo make server` & `cargo make watch` - https://github.com/seed-rs/seed-quickstart#3-prepare-your-project-for-work ; `cargo make start` is a little bit simpler but becomes annoying quickly because of restarts.
> 

Fixed on main site (getting started page). Leaving the `make start` on the main page because its a bit quicker.

> * [ ]  4) https://seed-style-hooks.netlify.app/getting_started -> `How it works`. It would be nice to include a super simple example for each a-h) point.

Fixed on main site.

> 
> * [ ]  5) App panics with too general error (`Option::unwrap`) when `presets` is missing and you want to use e.g. `seed_colors::Red::No4`.

Fixed on main site.
> 
> * [ ]  6) I was a bit confused because of many (glob) imports - which libraries are imported and required and where I can find `seed_style`, `seed_colors`, `theme`, etc.. and what rules should I respect to not cause panics in runtime. Especially when rust-analyzer doesn't work. Perhaps just a beginner problem, but I had an impression that there are too many moving parts.

TODO 

> 
> * [ ]  7) Why we use `GlobalStyle::default()` and not `GlobalStyle::new()`? My first idea was that `GlobalStyle::default` has some predefined _default_ styles.
> 
>     * [ ]  8) I don't have a clear idea how should I use the code below / how it really works:
>       ```rust
>           use_themes(
>               || vec![theme::theme(), seed_style::style_presets()],
>               || {
>                   global_styles::use_themed_styles();
>                   themed_view(model)
>               },
>           )
>       ```
>       
>       
>       
>       * Why is there a `vec!`? What can (or should) I write into it? Should I keep ordering? What ordering?
>       * When it can panics?
>       * What does `use_themed_styles` mean?
>       * What can I (should) write into the block in the second argument?
>       * Why `global_styles` are used "after" `theme::theme()`? - the code isn't corresponding with my intuition (theme styles should override global style and therefore they should be instantiated after global styles)
>       * etc.
> 
> * [ ]  9) I'm not sure how `GlobalStyle` works. My intuition tells me that it should be something like singleton, however there are two `GlobalStyle::default()` in `global_styles.rs` (in Basic quickstart). And what `.activate_init_styles` and `.activate_styles` do? (And these two calls look a bit error-prone - I can imagine I'll forget to call them - maybe we should add at least tag `#[must_use]` somewhere.

Fixed, now all of this is handled in the Seed `init`.

> 
> * [ ]  10) `s().` vs `S.`. It's probably the most important thing to resolve before we can integrate Seed Style into Seed. So.. why `s().` and not `S.`?


TODO

> 
> * [ ]  11) Why I had to call `.breakpoint_scale([600, 960, 1280, 1920])` (https://seed-style-hooks.netlify.app/responsive_styling)?
> 
> * [ ]  12)
>       ```rust
>            Msg::WindowResized => {
>               conditionally_skip_rendering::<Breakpoint, _, _, _>(|| default_bp_theme(), orders)
>           }
>       ```
>       
>       
>   looks pretty cumbersome and I assume that `conditionally_skip_rendering` is imported by glob so it pollutes API surface and I don't have an idea where it's defined. Possible alternative - the first idea:
>       ```rust
>           Msg::WindowResized => {
>               if not(S::breakpoint(default_bp_theme()).has_changed()) { orders.skip(); }
>           }
>       ```
> 

TODO


> * [ ]  13) It would be nice to see the code snippet corresponding to the example - https://seed-style-hooks.netlify.app/layout
> 

TODO


> * [ ]  14) There are missing links to the repo and for feedback - something like the footer in https://seed-rs.org/. Also I don't see a link to seed-rs.org.

TODO
